merino/providers/wikipedia/backends/elastic.py

```
@@ -0,0 +1,72 @@
+"""The Elasticsearch backend for Dynamic Wikipedia."""
+import logging
+from typing import Any, Final
+from urllib.parse import quote
+
+from elasticsearch import AsyncElasticsearch
+
+from merino.config import settings
+
+# The Index ID in Elasticsearch cluster.
+INDEX_ID: Final[str] = settings.providers.wikipedia.es_index
+SUGGEST_ID: Final[str] = "suggest-on-title"
+TIMEOUT_MS: Final[str] = f"{settings.providers.wikipedia.es_request_timeout_ms}ms"
+MAX_SUGGESTIONS: Final[int] = settings.providers.wikipedia.es_max_suggestions
+
+logger = logging.getLogger(__name__)
+
+
+class ElasticBackend:
+    """The client that works with the Elasticsearch backend."""
+
+    client: AsyncElasticsearch
+
+    def __init__(self, *, cloud_id: str, user: str, password: str) -> None:
+        """Initialize."""
+        self.client = AsyncElasticsearch(cloud_id=cloud_id, basic_auth=(user, password))
+
+    async def shutdown(self) -> None:
+        """Shut down the connection to the ES cluster."""
+        await self.client.close()
+
+    async def search(self, q: str) -> list[dict[str, Any]]:
+        """Search suggestions for a given query from the ES cluster."""
+        suggest = {
+            SUGGEST_ID: {
+                "prefix": q,
+                "completion": {
+                    "field": "suggest",
+                    "size": MAX_SUGGESTIONS,
+                },
+            }
+        }
+
+        try:
+            res = await self.client.search(
+                index=INDEX_ID,
+                suggest=suggest,
+                timeout=TIMEOUT_MS,
+                source_includes=["title"],
+            )
+        except Exception as e:
+            logger.warning(f"Failed to search from ES: {e}")
+            return []
+
+        if "suggest" in res:
+            return [
+                self.build_article(doc)
+                for doc in res["suggest"][SUGGEST_ID][0]["options"]
```

Hmm... the nested `res` data access scares me. We only check that "suggest" is in the response, but not the other options. There is potential for any of the subsequent access to fail too... so maybe we should wrap this with a `try` instead?

like:

```
try:
    articles = res["suggest"][SUGGEST_ID][0]["options"]
except KeyError:
    return []

return [ self.build_article(doc) for doc in articles ]
```

or something?

---

Yeah, Elastic really likes using deeply nested objects in the response 🤷‍♂️

Per its API doc, if `suggest` is present in the response, meaning it has found some hits for this query, that's why I only check that in the if statement. As for other lookups, I opted to not be super defensive because that could silently ignore those API change failures in the future. When that happens at some point, this will fail fast and loudly so that our tests or even Sentry will capture it right away.

Does that make sense? If so, I can add some comments here to clarify.

---

> I opted to not be super defensive because that could silently ignore those API change failures in the future. 

Good point! Let's leave it less defensive.